_posts/2023-11-01-#28368.md

```diff
@@ -7,18 +7,62 @@ authors: [ismaelsadeeq]
 components: ["tx fees and policy"]
 host: ismaelsadeeq
 status: upcoming
-commit:
+commit: 79bcc5ca0679daf6e57fc4d7ce2244262a7cfd13
 ---
 
-_Notes and questions to follow soon!_
 
-<!-- TODO: Before meeting, add notes and questions
 ## Notes
 
+* The [`CBlockPolicyEstimator`](https://github.com/bitcoin/bitcoin/blob/ab2f531b785c3c17746ac9766a6db5b6d26677e8/src/policy/fees.h#L146C13-L146C13) is used to estimate fees based on a confirmation target. We have two RPCs, [`estimaterawfees`](https://github.com/bitcoin/bitcoin/blob/ab2f531b785c3c17746ac9766a6db5b6d26677e8/src/rpc/fees.cpp#L100) and [`estimatesmartfees`](https://github.com/bitcoin/bitcoin/blob/ab2f531b785c3c17746ac9766a6db5b6d26677e8/src/rpc/fees.cpp#L29) used to obtain a fee estimate for a given confirmation target, it is also heavily used in the [wallet](https://github.com/bitcoin/bitcoin/blob/ab2f531b785c3c17746ac9766a6db5b6d26677e8/src/wallet/fees.cpp#L29).
+
+* Before this PR [`CTxMempool`](https://github.com/bitcoin/bitcoin/blob/22fa1f4702e9a60e141f545f8d09704deca34b22/src/txmempool.h#L301) owned a [`CBlockPolicyEstimator`](https://github.com/bitcoin/bitcoin/blob/ab2f531b785c3c17746ac9766a6db5b6d26677e8/src/policy/fees.h#L146C13-L146C13) as one of its members such that whenever a transaction is added or removed from the mempool in [`addUnchecked`](https://github.com/bitcoin/bitcoin/blob/ab2f531b785c3c17746ac9766a6db5b6d26677e8/src/txmempool.h#L475) and [`removedUnchecked`](https://github.com/bitcoin/bitcoin/blob/ab2f531b785c3c17746ac9766a6db5b6d26677e8/src/txmempool.h#L787) methods we also update the fee estimator synchronously.
+
+* Similarly, whenever a new block is connected, the [`removeForBlock`](https://github.com/bitcoin/bitcoin/blob/ab2f531b785c3c17746ac9766a6db5b6d26677e8/src/txmempool.h#L487) method removes the block transactions that are in the mempool together with its ancestors. This also triggers an update to the fee estimator synchronously.
+ 
+* Since `processBlock` happens within [`removeForBlock`](https://github.com/bitcoin/bitcoin/blob/ab2f531b785c3c17746ac9766a6db5b6d26677e8/src/txmempool.h#L487), these fee estimator update blocks [`ConnectTip`](https://github.com/bitcoin/bitcoin/blob/ab2f531b785c3c17746ac9766a6db5b6d26677e8/src/validation.h#L732C13-L732C13). Adding more steps to the fee estimator during this update can slow down block processing even further.
+
+* The [`CValidationInterface`](https://github.com/bitcoin/bitcoin/blob/ab2f531b785c3c17746ac9766a6db5b6d26677e8/src/validationinterface.h#L77) is a class any subscriber can implement to register for events and receive notifications when the events are fired.
```

Comment:
Hm I think there is some conflation between registering with `CMainSignals` and implementing a method of `CValidationInterface`. I'd probably describe it as:

Subclasses of `CValidationInterface` implement its methods to specify what should be done in response to the corresponding event, and register these callbacks with `CMainSignals`. Validation and mempool use `CMainSignals` to "notify" subscribers of these events. The "notification" may entail executing a callback immediately or adding it to a queue of callbacks to be executed in the background. These classes can be called clients or subscribers of the `CValidationInterface`.

Reply:
Thanks, I have updated it to.
Subclasses of [CValidationInterface](https://github.com/bitcoin/bitcoin/blob/9c30f5ef9d6712539cb8dc0f52b3a2206a25e074/src/validationinterface.h#L77) can implement its methods, and specify the instructions to be executed in response to the corresponding event. and register the subclass with [CMainSignals](https://github.com/bitcoin/bitcoin/blob/9c30f5ef9d6712539cb8dc0f52b3a2206a25e074/src/validationinterface.h#L182C7-L182C19) using [RegisterValidationInterface](https://github.com/bitcoin/bitcoin/blob/9c30f5ef9d6712539cb8dc0f52b3a2206a25e074/src/validationinterface.h#L26). Validation and mempool use [CMainSignals](https://github.com/bitcoin/bitcoin/blob/9c30f5ef9d6712539cb8dc0f52b3a2206a25e074/src/validationinterface.h#L182C7-L182C19) to "notify" subscribers of these events. The "notification" may entail executing a callback immediately or adding it to a queue of callbacks to be executed in the background.